pkg/util/buffer.go

```diff
@@ -0,0 +1,201 @@
+package util
+
+import (
+	"container/list"
+	"context"
+	"io"
+	"slices"
+	"sync"
+)
+
+// StreamBuffer is an in-memory buffer that can simultaneously be written to
+// by a single writer, and read from by multiple readers, such that all readers
+// see the same data.
+//
+// To read from the buffer, call NewStream(). This returns a channel that
+// streams a copy of the buffer's contents. If there is already data
+// accumulated in the buffer, the channel will begin to read it immediately,
+// then continue to receive any new data written to the buffer until the buffer
+// is closed. Each channel returned by NewStream() receives a copy of the
+// buffer's full contents, regardless of when it was created.
+//
+// StreamBuffer implements io.WriteCloser. When Close() is called, all stream
+// channels are closed and any subsequent writes to the buffer return
+// io.ErrClosedPipe. After the buffer is closed, calls to NewStream return a
+// channel that will receive the full contents of the buffer, then be closed.
+//
+// Care should be taken to ensure that the buffer is always closed when no
+// more writes are expected, because the stream channels are expected to be
+// read from until they are closed. Leaving the buffer open will cause all
+// readers to block indefinitely.
+type StreamBuffer struct {
+	mu     sync.Mutex
+	chunks *list.List
+	closed bool
+}
+
+const (
+	maxChunkSize = 4 * 1024
+)
+
+var _ io.WriteCloser = (*StreamBuffer)(nil)
+
+func NewStreamBuffer() *StreamBuffer {
+	buf := &StreamBuffer{
+		chunks: list.New(),
+	}
+	buf.chunks.PushBack(newChunk())
+	return buf
+}
+
+func (b *StreamBuffer) Write(p []byte) (n int, err error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.closed {
+		return 0, io.ErrClosedPipe
+	}
+
+	tail := b.acquireLastChunkLocked()
+	chunk := tail.Value.(*chunk)
+	chunk.Append(p)
+
+	return len(p), nil
+}
+
+func (b *StreamBuffer) Close() error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.closed {
+		return nil
+	}
+	b.closed = true
+	b.chunks.Back().Value.(*chunk).Seal()
+	return nil
+}
+
+func (b *StreamBuffer) NewStream(ctx context.Context) <-chan []byte {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	mu := sync.Mutex{}
+	var done bool
+	rc := make(chan []byte)
+	closeOnce := sync.OnceFunc(func() {
+		mu.Lock()
+		defer mu.Unlock()
+		done = true
+		close(rc)
+	})
+
+	go func() {
+		defer closeOnce()
+		for elem := b.headChunk(); elem != nil; elem = b.nextChunk(elem) {
+			off := 0
+			c := elem.Value.(*chunk)
+			for {
+				bytes, more := c.Next(off)
```

Comment:
This is forced to block until a new chunk is added to the output or the output is closed, even if the client is long gone. Condvars and contexts are really awkward to mix.

Reply:
Agreed - though fixable, it's still kinda awkward. :+1:

Reply:
Pushed an updated implementation - I think this one is much improved, also added some benchmark tests.